app/lib/view-transition.tsx

```diff
@@ -0,0 +1,290 @@
+    );
+  }
+}
+
+function removeAnimationDirectionStyles() {
+  document.documentElement.style.removeProperty('--direction-out');
+  document.documentElement.style.removeProperty('--direction-in');
+  document.documentElement.style.removeProperty('--view-transition-in-z-index');
+  document.documentElement.style.removeProperty(
+    '--view-transition-out-z-index',
+  );
+}
+
+type ViewTransitionState = {
+  transitionDirection: TransitionDirection;
+  transitionType: TransitionType;
+  transitionStyle: TransitionStyle;
+  previousPath: string; // so we can go back to the previous page
+};
+
+// location.state might be defined but set by someone else, so we need to check it
+function validateViewTransitionState(state: ViewTransitionState | null) {
+  if (
+    typeof state === 'object' &&
+    state !== null &&
+    'transitionDirection' in state &&
+    'transitionType' in state &&
+    'transitionStyle' in state &&
+    'previousPath' in state
+  ) {
+    return state as ViewTransitionState;
+  }
+
+  return null;
+}
+
+/**
+ * Applies the animation direction styles based on the navigation state.
+ * Must be used in the root component of the app.
+ */
+export function useViewTransitionEffect() {
+  const navigation = useNavigation();
+
+  useEffect(() => {
+    if (navigation.state === 'loading') {
+      const state = validateViewTransitionState(navigation.location.state);
+      if (!state) {
+        return removeAnimationDirectionStyles();
+      }
+
+      console.error('STATE', state);
+
+      const transitionDirection: TransitionDirection =
+        state.transitionDirection;
+      const transitionType: TransitionType = state.transitionType;
+      const transitionStyle: TransitionStyle = state.transitionStyle;
+
+      applyAnimationDirectionStyles(
+        transitionDirection,
+        transitionType,
+        transitionStyle,
+      );
+    }
+  }, [navigation]);
+}
+
+type ViewTransitionCommonProps = {
+  back?: boolean;
+  style?: TransitionStyle;
+} & (
+  | {
+      back: true;
+      direction?: never;
+      type?: never;
+      to?: never;
+    }
+  | {
+      back?: false;
+      direction: TransitionDirection;
+      type: TransitionType;
+      to: string;
+    }
+);
+
+type ViewTransitionLinkProps = ViewTransitionCommonProps & {
+  as?: typeof Link;
+} & Omit<React.ComponentProps<typeof Link>, 'to'>;
+
+type ViewTransitionNavLinkProps = ViewTransitionCommonProps & {
+  as: typeof NavLink;
+} & Omit<React.ComponentProps<typeof NavLink>, 'to'>;
+
+/**
+ * A wrapper around Link/NavLink that when used will animate the page transitions.
+ *
+ * Default is to prefetch the link when it is rendered to optimize the mobile experience,
+ * but this can be overridden by setting the prefetch prop.
+ *
+ * Special case: when back=true, navigates to previous path with opposite animation.
+ *
+ * @param props - direction: 'left' | 'right' | 'up' | 'down', type: 'open' | 'close', style: 'static' | 'dynamic'
```

Comment:
actually I am thinking if this api can be further simplified. so what we want is to tell it which animation to do and which view needs to animate so we don't need three props for this. open/close type only matters for dynamic  and that is a sign it can be simplified

What do you think about something like this instead https://github.com/MakePrisms/boardwalkcash/pull/217 ? It only has two params now to control the transition and the api describes the intent better imo

Reply:
Yea that's much better, merged #217